DuckDuckGo/Subscription/PrivacyProDataReporting.swift

```diff
@@ -0,0 +1,310 @@
+
+    private lazy var includedOrigins = config.settings(for: .additionalCampaignPixelParams)[Constants.includedOrigins] as? [String]
+
+    private let config: PrivacyConfiguration
+    private let variantManager: VariantManager
+    private let userDefaults: UserDefaults
+    private let emailManager: EmailManager
+    private let tutorialSettings: TutorialSettings
+    private let appSettings: AppSettings
+    private let statisticsStore: StatisticsStore
+    private let secureVault: (any AutofillSecureVault)?
+    private var syncService: DDGSyncing?
+    private var tabsModel: TabsModel?
+    private let dateGenerator: () -> Date
+
+    init(config: PrivacyConfiguration = ContentBlocking.shared.privacyConfigurationManager.privacyConfig,
+         variantManager: VariantManager = DefaultVariantManager(),
+         userDefaults: UserDefaults = .app,
+         emailManager: EmailManager = EmailManager(),
+         tutorialSettings: TutorialSettings = DefaultTutorialSettings(),
+         appSettings: AppSettings = AppDependencyProvider.shared.appSettings,
+         statisticsStore: StatisticsStore = StatisticsUserDefaults(),
+         secureVault: (any AutofillSecureVault)? = try? AutofillSecureVaultFactory.makeVault(reporter: SecureVaultReporter()),
+         syncService: DDGSyncing? = nil,
+         tabsModel: TabsModel? = nil,
+         dateGenerator: @escaping () -> Date = Date.init) {
+        self.config = config
+        self.variantManager = variantManager
+        self.userDefaults = userDefaults
+        self.emailManager = emailManager
+        self.tutorialSettings = tutorialSettings
+        self.appSettings = appSettings
+        self.statisticsStore = statisticsStore
+        self.secureVault = secureVault
+        self.syncService = syncService
+        self.tabsModel = tabsModel
+        self.dateGenerator = dateGenerator
+    }
+
+    func injectSyncService(_ service: DDGSync) {
+        syncService = service
+    }
+
+    func injectTabsModel(_ model: TabsModel) {
+        tabsModel = model
+    }
+
+    private var isReady: Bool {
+        syncService != nil && tabsModel != nil
+    }
+
+    /// Collect a randomized subset of parameters iff the Privacy Pro impression/conversion pixels
+    /// or the Origin Attribution subscription pixel are being fired
+    func randomizedParameters(for useCase: UseCase) -> [String: String] {
+        switch useCase {
+        case .messageID(let messageID):
+            guard let includedOrigins, includedOrigins.contains(messageID) else { return [:] }
+        case .origin(let origin):
+            guard let includedOrigins, let origin, includedOrigins.contains(origin) else { return [:] }
+        case .debug:
+            break
+        }
+
+        /// Wait for all the injected dependencies to be available
+        if !isReady && !ProcessInfo().arguments.contains("testing") {
+            return randomizedParameters(for: useCase)
+        }
```

Comment:
We run into this easily when the app launches and a remote message with the matching ID is shown.

Reply:
@quanganhdo testing using the provided configs worked as expected, however when I was regression testing with updated config id + messages, I kept ending up in an endless loop here. I was re-launching the app with an updated message config while on the new tab page with a remote message already showing (it looks like syncService is the one never being initialised)

<img width="1585" alt="image" src="https://github.com/user-attachments/assets/85736d0b-3d84-46d5-8f69-dd879d175e38">